pkg/hub/taint/controller.go

```diff
@@ -0,0 +1,95 @@
+package taint
+
+import (
+	"context"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+	clientset "open-cluster-management.io/api/client/cluster/clientset/versioned"
+	informerv1 "open-cluster-management.io/api/client/cluster/informers/externalversions/cluster/v1"
+	listerv1 "open-cluster-management.io/api/client/cluster/listers/cluster/v1"
+	v1 "open-cluster-management.io/api/cluster/v1"
+	"open-cluster-management.io/registration/pkg/helpers"
+)
+
+var (
+	UnavailableTaint = v1.Taint{
+		Key:    v1.ManagedClusterTaintUnavailable,
+		Effect: v1.TaintEffectNoSelect,
+	}
+
+	UnreachableTaint = v1.Taint{
+		Key:    v1.ManagedClusterTaintUnreachable,
+		Effect: v1.TaintEffectNoSelect,
+	}
+)
+
+// taintController
+type taintController struct {
+	clusterClient clientset.Interface
+	clusterLister listerv1.ManagedClusterLister
+	eventRecorder events.Recorder
+}
+
+// NewTaintController creates a new taint controller
+func NewTaintController(
+	clusterClient clientset.Interface,
+	clusterInformer informerv1.ManagedClusterInformer,
+	recorder events.Recorder) factory.Controller {
+	c := &taintController{
+		clusterClient: clusterClient,
+		clusterLister: clusterInformer.Lister(),
+		eventRecorder: recorder.WithComponentSuffix("taint-controller"),
+	}
+	return factory.New().
+		WithInformersQueueKeyFunc(func(obj runtime.Object) string {
+			accessor, _ := meta.Accessor(obj)
+			return accessor.GetName()
+		}, clusterInformer.Informer()).
+		WithSync(c.sync).
+		ToController("taintController", recorder)
+}
+
+func (c *taintController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	managedClusterName := syncCtx.QueueKey()
+	klog.V(4).Infof("Reconciling ManagedCluster %s", managedClusterName)
+	managedCluster, err := c.clusterLister.Get(managedClusterName)
+	if errors.IsNotFound(err) {
+		// Spoke cluster not found, could have been deleted, do nothing.
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	managedCluster = managedCluster.DeepCopy()
+	newTaints := managedCluster.Spec.Taints
+	cond := managedCluster.Status.Conditions
+
+	switch {
+	case meta.IsStatusConditionPresentAndEqual(cond, v1.ManagedClusterConditionAvailable, metav1.ConditionUnknown) ||
```

Comment:
I think you could

```go
cond := meta.FindStatusCondition(cond, v1.ManagedClusterConditionAvailable)
```

in case, maybe it could be

```
updated = helpers.RemoveTaints(&newTaints, UnavailableTaint)
updated = helpers.AddTaints(&newTaints, UnreachableTaint) && updated
```

then case based on cond.Status. Note that IsStatusConditionFalse always call FindStatusCondition. And since your removeTaints, addTatins return updated. I think you can do 

```
if updated
```
finally.

I think we are almost done. Thanks!